shared/src/main/scala/kinesis4cats/producer/fs2/FS2Producer.scala

```diff
@@ -111,33 +112,37 @@ abstract class FS2Producer[F[_], PutReq, PutRes](implicit
         .evalMap { x =>
           val c = ctx.addEncoded("batchSize", x.size)
           x.toNel
-            .fold(F.unit) { records =>
+            .fold(F.unit) { x =>
+              val records = x.map(_._1)
+              val deferreds = x.map(_._2)
               for {
                 _ <- logger.debug(c.context)(
                   "Received batch to process"
                 )
-                _ <- underlying
-                  .put(records)
-                  .flatMap(callback)
+                _ <- F.guaranteeCase(
+                  underlying
+                    .put(records)
+                    .flatMap(callback)
+                ) {
+                  case Canceled() | Succeeded(_) =>
+                    deferreds.traverse_(_.complete(F.unit))
+                  case Errored(e) =>
+                   val failedF = F.raiseError(e)
+                    deferreds.traverse_(d => d.complete(failedF))
```

Comment:
I think `underlying.put` gives back enough info, to be a bit more precise here (and only succeed records that indeed were successful). 
E.g
```
case class PutRecordsOutput(records: List[PutRecordsResultEntry], failedRecordCount: Option[NonNegativeIntegerObject] = None, encryptionType: Option[EncryptionType] = None)
```

Reply:
So the short answer is, it depends. But in more detail about the Producer interface.

There is the concept of a PutRecords _failure_ in which some or all of the records failed to produce with a given error code. This will result in a successful F for `put` but with a response that shows failed records. 

The F can also fail, but only if the PutRecords API fails as well. This is an exception, but also a much rarer occurrence. 

The Producer can be configured to retry both of these events with a RetryPolicy. 

If the retries are exhausted, the Producer can be configured to [raise an error](https://github.com/etspaceman/kinesis4cats/blob/main/shared/src/main/scala/kinesis4cats/producer/Producer.scala#L224) in the event of failed records being detected. By default, this is false (to match the Kinesis client behavior). Of course, if an exception is raised by the Kinesis client itself on the final retry, then that will be raised as well.

So with all that in mind, this routine depends on how the Producer is configured. If the producer is configured to raise errors after its retry loop is exhausted, then the inner F is there to capture that failure. 

I could certainly change this to be `F[F[Producer.Result]]` in the succeeded case, as that is probably useful for designing callbacks. Though I'm not sure how I'd capture the Cancelled case.